### PR TITLE
Fix #1643 : Fully qualified model name + Unique ShortID generation for DTO Model to avoid collision issues.

### LINKED ISSUE #1643: Bug: `500: Two different schemas with the title app.domain.authors.Author have been defined.`
### Description

Can be reproduced  by upgrading  pg-redis-docker to alpha6.
Reproduced example included -  https://github.com/v3ss0n/starlite-pg-redis-docker/tree/bug-alpha6

### URL to code causing the issue

http://localhost:8000/schema/swagger

### MCVE

```python
https://github.com/v3ss0n/starlite-pg-redis-docker/tree/bug-alpha6
```

### Steps to reproduce

```bash
1. Upgrade To Starlite Alpha6
2. Go to http://localhost:8000/schema/swagger
3. See error
```

### Screenshots

```bash
"![SCREENSHOT_DESCRIPTION](SCREENSHOT_LINK.png)"
```

### Logs

```
first: {"properties":{"id":{"type":"string","format":"uuid","description":"Any UUID string"},"name":{"type":"string"},"dob":{"type":"string","format":"date"},"country_id":{"oneOf":[{"type":"null"},{"type":"string","format":"uuid","description":"Any UUID string"}]}},"type":"object","required":["dob","name"],"title":"app.domain.authors.Author"}
starlite-pg-redis-docker-app-1    | second: {"properties":{"created":{"type":"string","format":"date-time"},"updated":{"type":"string","format":"date-time"},"id":{"type":"string","format":"uuid","description":"Any UUID string"},"name":{"type":"string"},"dob":{"type":"string","format":"date"},"country_id":{"oneOf":[{"type":"null"},{"type":"string","format":"uuid","description":"Any UUID string"}]},"nationality":{"oneOf":[{"type":"null"},{"$ref":"#/components/schemas/app.domain.authors.Author.nationality_0"}]}},"type":"object","required":["dob","name"],"title":"app.domain.authors.Author"}

if schema.title and schema.type in (OpenAPIType.OBJECT, OpenAPIType.ARRAY):
--
782 | if schema.title in schemas and hash(schemas[schema.title]) != hash(schema):
783 | raise ImproperlyConfiguredException(
784 | f"Two different schemas with the title {schema.title} have been defined.\n\n"
785 | f"first: {encode_json(schemas[schema.title].to_schema()).decode()}\n"
786 | f"second: {encode_json(schema.to_schema()).decode()}\n\n"
787 | f"To fix this issue, either rename the base classes from which these titles are derived or manually"
788 | f"set a 'title' kwarg in the route handler."
789 | )
790 | schemas[schema.title] = schema
```

### Litestar Version

litestar = { version = "==2.0.0a6" , extras = ["redis", "cli"] }

### Platform

- [X] Linux
- [ ] Mac
- [ ] Windows
- [ ] Other (Please specify in the description above)

### PR DESCRIPTION
**Pull request by v3ss0n**
### Pull Request Checklist

[//]: # "Please review the [Litestar contribution guidelines](https://github.com/litestar-org/litestar/blob/main/CONTRIBUTING.rst) for this repository."

- [x] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

### Description

[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

Bug fix for #1643  - where it cause duplicated DTOModel names collision. Attempted to fixed by adding a 12 char unique id to fully qualified name of model.

This PR adds two new test cases to ensure that the generated model names and schema names are unique

- First  test case to ensure that the model names generated by the backend are unique to avoid naming collisions between different parts of the code.
- Also next text case check that there are exactly two unique schemas. and that all schema names start with the modle name

### Close Issue(s)
  
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

Fixes #1643

### CODE REVIEW

**Review by peterschutt — COMMENTED** [review]
Hey @v3ss0n - thanks for taking this one on!

When we were talking in discord, I was only referring to the msgspec backend type, however there is also the pydantic backend, so perhaps we can create a utility method on `AbstractDTOBackend` that accepts `unique_name` as an argument and returns that with the short uid appended to it. We can then call that method from within `create_transfer_model_type` on either backend implementation.

Also, tests please:)

Thanks again!

**Comment by v3ss0n** [comment]
Yeah i saw other back-end uses it , i was about to ask that. Have to test all those around.
Is that ok to put it in AbstractDTOBackend ?  or should i add on all of those?

**Comment by v3ss0n** [comment]
```python
@pytest.mark.parametrize("backend_type", [MsgspecDTOBackend, PydanticDTOBackend])
def test_backend_model_name_uniqueness(backend_type: type[AbstractDTOBackend], backend_context: BackendContext) -> None:
    backend = backend_type(backend_context)
    unique_names: set = set()
    fd: FieldDefinitionsType = FieldDefinition(
        name="a",
        default=Empty,
        parsed_type=ParsedType(int),
        default_factory=None,
        dto_field=None,
        unique_model_name="some_module.SomeModel",
    )
    for i in range(100):
        model_class = backend.create_transfer_model_type("some_module.SomeModel", fd)
        model_name = model_class.unique_name
        assert model_name not in unique_names
        unique_names.add(model_name)
   ```
   
   I am trying like this , i don't know how to get unique_name back from the transfer_model .

**Comment by v3ss0n** [comment]
sync with main , resolved conflict at poetry.lock , its ready for review.

File: tests/dto/factory/test_integration.py [hunk 1]
```diff
@@ -176,3 +176,26 @@ def handler(data: DTOData[User]) -> User:
...
+    read_dto = DataclassDTO[Bar]
+
+    @post(dto=write_dto, return_dto=read_dto, signature_namespace={"Bar": Bar})
+    def handler(data: Bar) -> Bar:
+        return data
+
+    with create_test_client(route_handlers=[handler], debug=True) as client:
+        response = client.get("/schema/openapi.json")
+        schemas = list(response.json()["components"]["schemas"].values())
+        assert len(set(schemas)) == 2
```
Comment (peterschutt) [inline tests/dto/factory/test_integration.py:197]:
```suggestion
        assert len(schemas) == 2
        assert schemas[0] != schemas[1]
```
This is where the tests are failing, and is due to `schemas` being a `list` of `dict`. Converting that list into a `set` won't work as anything that goes into a `set` needs to be hashable, and `dict`'s are not hashable.

**Review by peterschutt — CHANGES_REQUESTED** [review]
Almost there, just need to get that test passing.

File: tests/dto/factory/test_integration.py (hunk 1 above)
Comment (v3ss0n) [reply tests/dto/factory/test_integration.py:197]: Thanks : This resulting in error unhashable type. my bad.

Comment (v3ss0n) [reply tests/dto/factory/test_integration.py]: tested , and it is passing .

**Comment by peterschutt** [comment]
Just tested this on the example app - the first 9 characters of each of those random strings is identical across all of the models which I think is due to the first part of uuid1 being a timestamp. This is why you needed to increase the size of the random string in 8e025cb6da77b5f2b5e5da710c750660d22a6b2b - because the first chars of a timestamp are too slow-moving to be reliably different.

I think we'd be better of just using something like `base64.urlsafe_b64encode(os.urandom(6))` for the random component.

Sorry I didn't pick this up earlier!

**Comment by provinzkraut** [comment]
> I think we'd be better of just using something like `base64.urlsafe_b64encode(os.urandom(6))`

Why not just use `secrets.token_urlsafe(6)` for this?

**Comment by peterschutt** [comment]
> > I think we'd be better of just using something like `base64.urlsafe_b64encode(os.urandom(6))`
> 
> Why not just use `secrets.token_urlsafe(6)` for this?

Perfect.

**Comment by v3ss0n** [comment]
> Just tested this on the example app - the first 9 characters of each of those random strings is identical across all of the models which I think is due to the first part of uuid1 being a timestamp. This is why you needed to increase the size of the random string in [8e025cb](https://github.com/litestar-org/litestar/commit/8e025cb6da77b5f2b5e5da710c750660d22a6b2b) - because the first chars of a timestamp are too slow-moving to be reliably different.
> 
> I think we'd be better of just using something like `base64.urlsafe_b64encode(os.urandom(6))` for the random component.
> 
> Sorry I didn't pick this up earlier!

I was thinking that too , with uuid1 looks like we need whole of it.
I am gonna try with secrets.token_urlsafe(6)

**Comment by v3ss0n** [comment]
secrets.token_urlsafe(6) have dashes and undercores , should thate be ok?

**Comment by provinzkraut** [comment]
> secrets.token_urlsafe(6) have dashes and undercores , should thate be ok?

Probably better not to. Try `secrets.token_hex`. That's just letters and numbers.

**Comment by v3ss0n** [comment]
> > secrets.token_urlsafe(6) have dashes and undercores , should thate be ok?
> 
> Probably better not to. Try `secrets.token_hex`. That's just letters and numbers.

thanks , this look better , going to raise to 8 chars to increase randomness.
```python
In [30]: secrets.token_hex(8)
Out[30]: '342e6bed9da5e8a5'
```

**Comment by v3ss0n** [comment]
changed to : secrets.token_hex(8) , tested by incresing loop to 100k , passed locally.

**Review by peterschutt — APPROVED** [review]

**Comment by peterschutt** [comment]
@all-contributors add @v3ss0n for code!

**Comment by v3ss0n** [comment]
Thanks a lot!!

**Pull request merged**